packages/vite/src/node/server/index.ts

```diff
@@ -700,14 +710,21 @@ export function resolveServerOptions(
   raw?: ServerOptions
 ): ResolvedServerOptions {
   const server = raw || {}
-  const fsServeRoot = normalizePath(
-    path.resolve(root, server.fsServe?.root || searchForWorkspaceRoot(root))
+  let allowDirs = server.fsServe?.allow
+
+  if (!allowDirs) {
+    allowDirs = [server.fsServe?.root || searchForWorkspaceRoot(root)]
+  }
+  allowDirs.push(CLIENT_DIR)
```

Comment:
Maybe we could detect if `CLIENT_DIR` is needed here before adding it to `allowDirs`, so we avoid showing two dirs in error messages in case people are using vite from node_modules. GitHub doesn't show me what I need for the suggestion:
```ts
function resolvedAllowDir(dir: string, root: string): string {
  return ensureLeadingSlash(normalizePath(path.resolve(root, dir)))
}

export function resolveServerOptions(
  root: string,
  raw?: ServerOptions
): ResolvedServerOptions {
  const server = raw || {}
  let allowDirs = server.fsServe?.allow

  if (!allowDirs) {
    allowDirs = [server.fsServe?.root || searchForWorkspaceRoot(root)]
  }

  allowDirs = allowDirs.map((i) => resolvedAllowDir(i, root))

  const resolvedClientDir = resolvedAllowDir(CLIENT_DIR, root)
  if (!allowDirs.some(i => resolvedClientDir.startsWith(i))) {
    allowDirs.push(resolvedClientDir)
  }

  server.fsServe = {
    // TODO: make strict by default
    strict: server.fsServe?.strict,
    allow: allowDirs
  }
  return server as ResolvedServerOptions
}
```